src/pip/_internal/operations/prepare.py

```diff
@@ -237,8 +238,23 @@ def unpack_url(
         unpack_vcs_link(link, location)
         return None
 
-    # If it's a url to a local directory
+    # Once out-of-tree-builds are no longer supported, could potentially
+    # replace the below condition with `assert not link.is_existing_dir`
+    # - unpack_url does not need to be called for in-tree-builds.
+    #
+    # As further cleanup, _copy_source_tree and accompanying tests can
+    # be removed.
     if link.is_existing_dir():
+        deprecated(
+            "A future pip version will change local packages to be built "
+            "in-place without first copying to a temporary directory. "
+            "We recommend you use --use-feature=in-tree-build to test "
+            "your packages with this new behavior before it becomes the "
+            "default.\n",
+            replacement=None,
+            gone_in=None,
```

Comment:
We should set a timeline for this -- let's go with pip 21.2 (Q2 2021)?

Reply:
Because this warning may go unnoticed in CI runs, and the change may be somewhat non-trivial where it actually causes issues, we may want a 3 steps deprecations:

1. warn on copytree, encourage users to enable the in-tree build feature (this PR)
2. error on copytree, allow users to use out-of-tree build with `--enable-legacy` (so people who did not notice still have a cycle to adapt)
3. remove copytree

Reply:
The general idea with the feature flags is to "allow early adoption *and* late adoption, while not creating busy work for folks who aren't affected".

Erroring out by default would create such busy work.

Reply:
@pradyunsg one of the show stoppers that led to the revert of #7882 was https://github.com/pypa/pip/issues/8165#issuecomment-624669107 (the change silently producing invalid build artifacts).

So my worry is that people will not notice the warning and wrong builds will happen silently when we flip the default.

Reply:
@sbidoul Am I correctly understanding the progression that you're suggesting - there will be a point (step 2) where running pip without any options will cause an error, and the only way to get pip to run will be to explicitly state whether or not you want in-place or out-of-tree builds?

That seems like an extremely heavy handed way of protecting a relatively small number of (admittedly quite serious) failure cases. I think at some point we have to acknowledge that it's the responsibility of the projects to spot and act on deprecation warnings, or deal with the consequences...

(On the other hand, if I've misunderstood the steps in your proposed process, please could you clarify?)

Reply:
A possibly (reduced) step 2 could be to just error-out when building wheels (i.e. not on `pip install local_dir`), as that's the case which is most likely to have wider repercussions as well as more likely to affect project maintainers than users.

Reply:
Another idea: could `pip` return a non-zero exit code if any deprecation warnings are emitted (perhaps with a flag to disable that behaviour)?

Might be a way to get more people to notice a deprecation warning in CI.

Reply:
@pfmoore I agree it is heavy handed. Maybe I worry to much after the revert and that would indeed create busy-work for a majority that would not be impacted by the change. Hard to tell as always... we at least have the indication that the world did not collapse during the couple of weeks 7882 was released. The alternative is good communication around that particular deprecation, which might have been the main missing part with the previous attempt.

@davidhewitt the "exit code on deprecation" idea is interesting and might be worth investigating for the future. I guess we'd need to do some kind of inventory of past and current deprecations to see how that would play out for them.

Reply:
I think allowing affected users affected to opt-out (when the default changes) would be sufficient to prevent this from being a showstopper TBH -- it changes "pip now breaks your workflow and your only option is to pin to an older version" to "pip has a new default that breaks my workflow. You can re-enable the old behavior, but it is now deprecated and due for removal soon.".

Notice that folks who aren't affected get the new behavior (which is good!) and anyone who is affected doesn't break immediately without an escape hatch (which is nice!). Affected users are also nudged to transition away because it's now a deprecated feature (which is also nice!). :)